src/passes/init.cc

```diff
@@ -135,22 +153,154 @@ namespace
           locals.end(),
           std::inserter(rhs_found, rhs_found.begin()));
 
-        if (lhs_found.empty() && rhs_found.empty())
+        // we want to give preference to "strict" assignment statements,
+        // that is those which have no variables on one side of the
+        // assignment.
+        std::size_t lhs_count = count_locals(lhs, locals);
+        std::size_t rhs_count = count_locals(rhs, locals);
+        bool strict_assign = lhs_count == 0 || rhs_count == 0;
+
+        // compiler statements will never be right-assign, so we can
+        // use this fact later to help resolve some ambiguities
+        bool compiler_stmt =
+          std::any_of(lhs_found.begin(), lhs_found.end(), [](auto& loc) {
+            std::string name = loc.str();
+            return name.starts_with("unify$") || name.starts_with("out$") ||
+              name.starts_with("value$");
+          });
+
+        for (auto& loc : lhs_found)
         {
-          continue;
+          if (init_stmt_for_local.contains(loc) && !strict_assign)
+          {
+            continue;
+          }
+
+          if (compiler_stmt)
+          {
+            init_stmt_for_local[loc] = {
+              i, strict_assign, lhs_found, std::set<Location>{}};
+          }
+          else
+          {
+            init_stmt_for_local[loc] = {i, strict_assign, lhs_found, rhs_found};
+          }
         }
 
-        for (auto& loc : lhs_found)
+        if (!compiler_stmt)
         {
-          locals.erase(loc);
+          for (auto& loc : rhs_found)
+          {
+            if (init_stmt_for_local.contains(loc) && !strict_assign)
+            {
+              continue;
+            }
+
+            init_stmt_for_local[loc] = {i, strict_assign, lhs_found, rhs_found};
+          }
         }
+      }
+    }
 
-        for (auto& loc : rhs_found)
+    // transform the init statements into LiteralInit nodes
+    std::vector<InitInfo> init_stmts;
+    std::transform(
+      init_stmt_for_local.begin(),
+      init_stmt_for_local.end(),
+      std::back_inserter(init_stmts),
+      [](auto& pair) { return pair.second; });
+
+    // sort the init statements so that statements with no local variables
+    // on one side come first.
+    std::sort(init_stmts.begin(), init_stmts.end(), [](auto& a, auto& b) {
```

Comment:
Is sorting sufficient here, or do you want a topological sort.

I guess I am worried about 
```
a = b
b = c
c = 1
```

Reply:
I tested that example and it does work, however, I think a topological sort is probably more principled. Good point.

Reply:
Updated code does a more principled sort as suggested, thanks again!